File: python solutions/toss_strange_coin.py
from functools import cache
from typing import List


class Solution:
    def probabilityOfHeads(self, prob: List[float], target: int) -> float:

        @cache
        def fn(i, k):
            if k < 0: return 0
            if i == len(prob): return k == 0
            return prob[i]*fn(i+1, k-1) + (1-prob[i])*fn(i+1, k)

        return fn(0, target)

File: ruby solutions/toss_strange_coin.rb
# @param {Float[]} prob
# @param {Integer} target
# @return {Float}
def probability_of_heads(prob, target)
    dp = [1] + [0] * target
    prob.each {
        |p|
        (target).downto(0).each { 
            |k| 
            dp[k] = (k > 0 ? dp[k - 1] : 0) * p + dp[k] * (1 - p) 
        }
    }
    dp[target]
end